File: medium/2095. Delete the Middle Node of a Linked List/2095. Delete the Middle Node of a Linked List.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return None

        if not head.next.next:
            head.next = None
            return head

        prev: ListNode = ListNode(0)
        slow: Optional[ListNode] = head
        fast: Optional[ListNode] = head

        while slow and fast and fast.next:
            prev.next = slow
            prev = prev.next
            slow = slow.next
            fast = fast.next.next

        assert slow is not None
        prev.next = slow.next
        return head

File: medium/2095. Delete the Middle Node of a Linked List/2095. Delete the Middle Node of a Linked List.go
package main

/**
 * Definition for singly-linked list.
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func deleteMiddle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}

	if head.Next.Next == nil {
		head.Next = nil
		return head
	}

	slow := head
	fast := head.Next.Next

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	slow.Next = slow.Next.Next
	return head
}
